chunker: test detection of incompatible metadata, fix NewObject on case-insensitive filesystem

ivandeex opened the pull request:
#### What is the purpose of this change?

Chunker was developed with a goal to work smoothly with most of 30+ rclone storage providers. As no single size fits all, chunker was engineered with versioning in mind. The main principles are:
- newer chunker/rclone versions accept and transparently upgrade metadata produced by _all_ previous versions
- if growing number of formats becomes hard to maintain, the ancient formats must be deprecated _explicitly_, at least one rclone release _in advance_
- chunker is smart enough to tell valid metadata from random garbage
- chunker can detect valid but unsupported metadata produced by future versions (vs garbage)
- upon detecting future metadata chunker must **refuse to modify** such file and explicitly ask user to **update** rclone

Hopefully this rather conservative approach will facilitate wider adoption of chunker/rclone. 

Development of a new metadata format by PR https://github.com/rclone/rclone/pull/4675#issuecomment-740206801 has discovered cases when chunker in rclone v1.50.0-v1.53.3 **fails to detect** composite objects produced and allows to inadvertently corrupt user data. This PR seeks to fix that and proposes a few provisions for compatiblity with future formats.

This PR is separate from #4675 due to different set of targeted rclone releases. I would like to merge it for v1.54 and also to backport it to **v1.53 stable**. I will help to rebase PR #4675 on top of this one.
rclone releases v1.50,v1.51 will **stay affected** by potential corruption. This PR will add a warning in the chunker documentation urging users of affected versions to upgrade.

@ncw Is that at all possible to backport to v1.52? 🙏🤗

The problem with metadata detection is that while chunker uses both directory scanning and metadata reading to detect composite objects, chunker tries to optimize out file reading as much as possible to save on storage API calls/charges. First, there is hardcoded metadata size limit, and files larger then that are not even considered. Next, chunker uses pattern of file names in a given directory to predict whether given set of files are parts of a composite object. Before this patch chunker required that there is at least one data chunk to start checking for a composite.

This PR extends the logic gently: if chunker cannot find data chunks accompanying metaobject but finds at least one potential control or temporary chunk, it marks found files as a _suspected_ composite object and marks its metaobject for _lazy_ reading later (too large metaobjects are thrown away as before). When rclone later tries a _concrete_ operation on the object, chunker performs postponed read and takes the final verdict: _is this compatible composite object, future metadata or just a garbage_? This approach allows to keep directory scan as fast as before but detect future chunked objects better.

Additionally, this PR splits single metadata size limit in two: _maxMetadataSizeWrite_ (255) is the current hard maximum metadata object produced by rclone, _maxMetadataSize_ (1023) is the new increased threshold for files considered by chunker for metadata scan. It is higher to make room for future metadata improvements. Nevertheless, as before, every increase of the _former_ limit _must_ increment the produced metadata version to prevent incorrect treatment and make user update ancient installations of rclone.

I am also hiding from command line a few chunker options which critically change chunker's notion of what a chunk is. I would like to prohibit them on command line completely if rclone provided such an option. As it's imporrible for now, chunker's user documentation provides users with appropriate warnings.

#### Was the change discussed in an issue or in the forum before?

Yes, in https://github.com/rclone/rclone/pull/4675#issuecomment-739639666 and https://github.com/rclone/rclone/pull/4675#issuecomment-740206801

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ivandeex commented:
I am including here a fix for the recent issue #4830 (NewObject on a case-insensitive FS).
The quick integration test suite now passes: https://github.com/ivandeex/rclone/runs/1650174935?check_suite_focus=true

ivandeex commented:
@ncw Please review

ivandeex commented:
@ncw Please set due date on the 1.54 milestone.

ncw commented:
> Chunker was developed with a goal to work smoothly with most of 30+ rclone storage providers. As no single size fits all, chunker was engineered with versioning in mind. The main principles are:
> 
> * newer chunker/rclone versions accept and transparently upgrade metadata produced by _all_ previous versions
> * if growing number of formats becomes hard to maintain, the ancient formats must be deprecated _explicitly_, at least one rclone release _in advance_
> * chunker is smart enough to tell valid metadata from random garbage
> * chunker can detect valid but unsupported metadata produced by future versions (vs garbage)
> * upon detecting future metadata chunker must **refuse to modify** such file and explicitly ask user to **update** rclone
> 
> Hopefully this rather conservative approach will facilitate wider adoption of chunker/rclone.

Seems reasonable. Though we might want to specify a deprecation period (1 year say) rather than tying it to the rclone release schedule.

> Development of a new metadata format by PR [#4675 (comment)](https://github.com/rclone/rclone/pull/4675#issuecomment-740206801) has discovered cases when chunker in rclone v1.50.0-v1.53.3 **fails to detect** composite objects produced and allows to inadvertently corrupt user data. This PR seeks to fix that and proposes a few provisions for compatiblity with future formats.

OK

> This PR is separate from #4675 due to different set of targeted rclone releases. I would like to merge it for v1.54 and also to backport it to **v1.53 stable**. I will help to rebase PR #4675 on top of this one.
> rclone releases v1.50,v1.51 will **stay affected** by potential corruption. This PR will add a warning in the chunker documentation urging users of affected versions to upgrade.

So merge this PR first the rebase #4675?

> @ncw Is that at all possible to backport to v1.52?

I think this patch is too big for a point release. I try to keep the point release patches really small as not to break stuff. So I think we should target this for 1.54 which I'll try to get out before the end of January 2021.

I don't normally update previous rclones once a new .0 has been released to try and keep the number of releases we need to maintain down.

> The problem with metadata detection is that while chunker uses both directory scanning and metadata reading to detect composite objects, chunker tries to optimize out file reading as much as possible to save on storage API calls/charges. First, there is hardcoded metadata size limit, and files larger then that are not even considered. Next, chunker uses pattern of file names in a given directory to predict whether given set of files are parts of a composite object. Before this patch chunker required that there is at least one data chunk to start checking for a composite.
> 
> This PR extends the logic gently: if chunker cannot find data chunks accompanying metaobject but finds at least one potential control or temporary chunk, it marks found files as a _suspected_ composite object and marks its metaobject for _lazy_ reading later (too large metaobjects are thrown away as before). When rclone later tries a _concrete_ operation on the object, chunker performs postponed read and takes the final verdict: _is this compatible composite object, future metadata or just a garbage_? This approach allows to keep directory scan as fast as before but detect future chunked objects better.
> 
> Additionally, this PR splits single metadata size limit in two: _maxMetadataSizeWrite_ (255) is the current hard maximum metadata object produced by rclone, _maxMetadataSize_ (1023) is the new increased threshold for files considered by chunker for metadata scan. It is higher to make room for future metadata improvements. Nevertheless, as before, every increase of the _former_ limit _must_ increment the produced metadata version to prevent incorrect treatment and make user update ancient installations of rclone.

OK

> I am also hiding from command line a few chunker options which critically change chunker's notion of what a chunk is. I would like to prohibit them on command line completely if rclone provided such an option. As it's imporrible for now, chunker's user documentation provides users with appropriate warnings.

That is the normal process for deprecation - Hide them from the command line and configurator, make them generate a warning (or error) then eventually remove. If you want them to generate an error then you'll need to do that by hand at the moment (in `NewFs`).

I'll review the code now

ncw commented on backend/chunker/chunker.go:
```diff
@@ -150,6 +150,7 @@ Normally should contain a ':' and a path, e.g. "myremote:path/to/dir",
 		}, {
 			Name:     "name_format",
 			Advanced: true,
+			Hide:     fs.OptionHideCommandLine,
```
You can `HideBoth` if you want `rclone config` to no longer offer them to users. They will carry on being read from existing config files.

ivandeex replied:
Yet I would like to have it offered for new remotes.

ncw commented on docs/content/chunker.md:
```diff
@@ -299,6 +299,9 @@ If wrapped remote is case insensitive, the chunker overlay will inherit
 that property (so you can't have a file called "Hello.doc" and "hello.doc"
 in the same directory).
 
+Chunker included in rclone releases up to `v1.53.2` can sometimes fail to
```
This should probably say "Chunker included in rclone releases prior to `v1.54.0` can..."

Note that we've released v1.53.3 already! I think we should target this for v1.54.0

ivandeex replied:
Yes, I intended the sequence of commits to be:
- merge it on master for v1.54.0
- cherry pick from master to v.1.53-stable for the next point release
- rebase PR 4675 on new master

ivandeex replied:
Do you mean that 1.53.4 release would be close to 1.54, making backport moot?

ncw reviewed (commented):
I've given this code a light review - looks very good :-)

See a few inline comments.

ivandeex commented:
> > * if growing number of formats becomes hard to maintain, the ancient formats must be deprecated _explicitly_, at least one rclone release _in advance_
> Seems reasonable. Though we might want to specify a deprecation period (1 year say) rather than tying it to the rclone release schedule.

I'm not sure about the schedule. I'm tying that with the `hard to maintain` condition.

> So merge this PR first the rebase #4675?

Yes
> I think this patch is too big for a point release. I try to keep the point release patches really small as not to break stuff. So I think we should target this for 1.54 which I'll try to get out before the end of January 2021.

That is sad. My intention was to reduce number of releases affected by the bug.
Would it be possible to backport to 1.53-stable the single commit "chunker: ensure that chunker rejects future metadata" alone, if I factor it to a separate PR?
(https://github.com/rclone/rclone/pull/4902/commits/71b23f9bb2ce77fb9f424900d8d20aad3aaccb49)

> I don't normally update previous rclones once a new .0 has been released to try and keep the number of releases we need to maintain down.

So backporting to 1.52.x is impossible as I expected.

ncw commented:
> > > * if growing number of formats becomes hard to maintain, the ancient formats must be deprecated _explicitly_, at least one rclone release _in advance_
> > >   Seems reasonable. Though we might want to specify a deprecation period (1 year say) rather than tying it to the rclone release schedule.
> 
> I'm not sure about the schedule. I'm tying that with the `hard to maintain` condition.

I'll leave the exact period up to you :-)

> > I think this patch is too big for a point release. I try to keep the point release patches really small as not to break stuff. So I think we should target this for 1.54 which I'll try to get out before the end of January 2021.
> 
> That is sad. My intention was to reduce number of releases affected by the bug.
> Would it be reasonable to backport to 1.53-stable the single commit "chunker: ensure that chunker rejects future metadata" alone, if I factored it as a separate PR?
> ([71b23f9](https://github.com/rclone/rclone/commit/71b23f9bb2ce77fb9f424900d8d20aad3aaccb49))

Yes that sounds reasonable. Howe about we do that first? I can accept it into 1.54 then we can backport it to 1.53. There are a couple of fixes to 1.53 we could push out too with it, and try and get that done in the next few days?
 
> > I don't normally update previous rclones once a new .0 has been released to try and keep the number of releases we need to maintain down.
> 
> So backporting to 1.52.x is impossible as I expected.

Not impossible, but I wonder of how much use? Do people follow the 1.52.x series when the 1.53.x series is available? I really don't know!

ivandeex commented:
- factored the core commit into PR #4917
- kept the commit here as this change was pre-rebased on top of the other one
- labeled #4917 as a candidate for 1.53.4, removed the mark here
- changed release number in documentation to 1.54
- done?

ncw reviewed (approved):
I think that looks great - please merge (after #4917) :-)
